### PR TITLE
fix: refactor APPLY_PATCH tool [ENG-1174]

### PR DESCRIPTION
**Pull request by abeatrix**
### Related Issue

**Issue:** https://linear.app/cline-bot/issue/ENG-1174

### Description

Replace separate file creation and editing tools with unified APPLY_PATCH tool for the native-gpt-5 model variant. This consolidation simplifies file operations through a single patch-based interface.

Changes:
- ~~Replace FILE_NEW and FILE_EDIT tools with APPLY_PATCH in config for gpt-5 with native tool calling~~ (will do this in follow-up)
- Disable EDITING_FILES system prompt section (no longer needed)
- Remove EDITING_FILES section from base template
- Refactor ApplyPatchHandler with improved architecture:
  - Extract patch parsing logic into PatchParser utility class
  - Extract file operations into FileProviderOperations utility
  - Extract path resolution into PathResolver utility
  - Add comprehensive error handling with DiffError types
  - Improve type safety with shared Patch types
- Add extensive test coverage for PatchParser including:
  - Edge cases (empty files, large files, unicode)
  - Error conditions (malformed patches, invalid operations)
  - Complex scenarios (multiple chunks, context matching)
- Export PATCH_MARKERS and BASH_WRAPPERS for reusability
- Add snapshots for native gpt-5 system prompt
- New UI for deleting file

This refactoring improves maintainability, testability, and provides a more robust patch application system for the GPT-5 model variant when native tool calling is enabled.

### Test Procedure

image

image

1. Start a new task using GPT-5-Codex model with Cline provider
2. Verify no error occurs when the apply patch tool is used
3. You can verify that the apply patch is being used by opening the "Cline wants to edit / create new file" dropdown to see if it's a patch 

image

### Type of Change

-   [ ] 🐛 Bug fix (non-breaking change which fixes an issue)
-   [ ] ✨ New feature (non-breaking change which adds functionality)
-   [ ] 💥 Breaking change (fix or feature that would cause existing functionality to not work as expected)
-   [x] ♻️ Refactor Changes
-   [ ] 💅 Cosmetic Changes
-   [ ] 📚 Documentation update
-   [ ] 🏃 Workflow Changes

### Pre-flight Checklist

-   [x] Changes are limited to a single feature, bugfix or chore (split larger changes into separate PRs)
-   [x] Tests are passing (`npm test`) and code is formatted and linted (`npm run format && npm run lint`)
-   [ ] I have created a changeset using `npm run changeset` (required for user-facing changes)
-   [x] I have reviewed [contributor guidelines](https://github.com/cline/cline/blob/main/CONTRIBUTING.md)

### Screenshots

Delete File UI :

https://github.com/user-attachments/assets/fa975403-567f-49d3-8b85-d72fc2f93baf

### Additional Notes

----

> [!IMPORTANT]
> Refactor `APPLY_PATCH` tool to consolidate file operations, improve error handling, and enhance test coverage, with updates to UI components for new tool types.
> 
>   - **Behavior**:
>     - Refactor `APPLY_PATCH` tool in `ApplyPatchHandler.ts` to consolidate file operations and improve error handling.
>     - Introduce `PatchParser`, `FileProviderOperations`, and `PathResolver` utility classes for better code organization.
>     - Add `DiffError` types for comprehensive error handling.
>   - **Testing**:
>     - Add extensive tests for `PatchParser` in `PatchParser.test.ts` covering edge cases, error conditions, and complex scenarios.
>   - **UI**:
>     - Update `ChatRow.tsx` and `TaskTimelineTooltip.tsx` to handle new `fileDeleted` tool type.
>     - Adjust button configurations in `buttonConfig.ts` for `fileDeleted` tool type.
>   - **Misc**:
>     - Export `PATCH_MARKERS` and `BASH_WRAPPERS` for reusability.
>     - Add snapshots for native gpt-5 system prompt.
> 
> <sup>This description was created by </sup>[Ellipsis](https://www.ellipsis.dev?ref=cline%2Fcline&utm_source=github&utm_medium=referral)<sup> for 9e3e31045de8fce1fa6edc14a317c9afed975174. You can [customize](https://app.ellipsis.dev/cline/settings/summaries) this summary. It will automatically update as commits are pushed.</sup>

### CODE REVIEW

File: src/shared/string.ts [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+	"\u00A0": " ", // NO-BREAK SPACE
+	"\u202F": " ", // NARROW NO-BREAK SPACE
+}
+
+/**
+ * Canonicalize a string by normalizing unicode punctuation and quotes
+ */
+export function canonicalize(s: string): string {
+	// First normalize unicode and punctuation
+	let normalized = s.normalize("NFC").replace(/./gu, (c) => PUNCT_EQUIV[c] ?? c)
```
[the remark this answers is not in the document]
Comment (abeatrix) [reply src/shared/string.ts:36]: think it's obvious that it's intentional

File: src/core/task/tools/handlers/ApplyPatchHandler.ts [hunk 2]
```diff
@@ -951,7 +634,7 @@ export class ApplyPatchHandler implements IFullyManagedTool {
 			return true
 		}
 
-		const fileCount = Object.keys(JSON.parse(messageStr).content.match(/\d+/)?.[0] || "0").length
+		const fileCount = Object.keys(JSON.parse(messageStr).content?.match(/\d+/)?.[0] || "0").length
```
[the remark this answers is not in the document]
Comment (abeatrix) [reply src/core/task/tools/handlers/ApplyPatchHandler.ts]: removed code

**Review by pashpashpash — APPROVED** [review]
This makes sense, but haven't had much time to review the code in detail. Stamping.

@0xToshii do you have a sec to also review this code?

**Comment by 0xToshii** [comment]
we should make a snap for the native gpt system prompt, unless i missed it

File: src/shared/string.ts [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+		.replace(/\\"/g, '"') // \" -> "
+
+	return normalized
+}
+
+/**
+ * Preserve the escaping style from original text when applying new text
+ * If original has \`, preserve that style in the replacement
+ */
+export function preserveEscaping(originalText: string, newText: string): string {
```
Comment (0xToshii) [inline src/shared/string.ts:54]: what was the history behind doing this? 

kinda edge case-y but are we concerned with examples like this?
```
		it("should not corrupt regex when simplifying escaped quote pattern", () => {
			// Original: Pattern with escaped literal quotes
			const original = "pattern = r'Input: \\\\\"\\\\w+\\\\\"'"

			// New: Simplified regex without literal quotes  
			const newText = "pattern = r'Input: \\w+'"  // \w+ captures words

			const result = preserveEscaping(original, newText)

			result.should.equal("pattern = r'Input: \\w+'")  // Should stay as \w+

			// Bug returns: "pattern = r'Input: \\\\w+'"  // Doubled to \\w+
		})
```

**Comment by abeatrix** [comment]
@0xToshii I've updated the PR to add the snapshots for native gpt-5 variant. Also keeping the current diff edit for native gpt-5 to make sure main is not affected. I will open a follow-up PR to add a new feature flag that allows us to use it for testing later.

File: src/core/task/tools/utils/FileProviderOperations.ts [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+	async modifyFile(path: string, content: string): Promise<FileOpsResult> {
+		this.provider.editType = "modify"
+		await this.provider.open(path)
+		await this.provider.update(content, true)
+		const result = await this.provider.saveChanges()
+		await this.provider.reset()
+		return result
+	}
+
+	async deleteFile(path: string): Promise<void> {
```
Comment (0xToshii) [inline src/core/task/tools/utils/FileProviderOperations.ts:35]: delete and move don't delete the original files rn, i think due to how provider is set up

also maybe we should consider better ux in the future for these 2. like the move being displayed as if the model was writing new code is kinda confusing

Comment (abeatrix) [reply src/core/task/tools/utils/FileProviderOperations.ts:35]: Yea i'm going to continue the UX work in a follow-up PR for the background edit: https://github.com/cline/cline/pull/7146

File: src/shared/string.ts (hunk 3 above)
Comment (abeatrix) [reply src/shared/string.ts:54]: This is only a failsafe in cases where the original content we sent to the LLM were escaped unintentionally between API calls. I've disable this check by default if this is a concern and we can do a A/B test later to determine if this is needed based on the edit failure rate later.
